shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java

```diff
@@ -1,178 +1,275 @@
+   * the Dart execution context. This communication is facilitate via 2 APIs:
+   *  - {@link BinaryMessenger}, which sends messages to Dart
+   *  - {@link PlatformMessageHandler}, which receives messages from Dart
+   */
   public void onAttachedToJNI() {
-    isAttachedToJni = true;
-    flutterJNI.setPlatformMessageHandler(this);
+    flutterJNI.setPlatformMessageHandler(messenger);
+    messenger.onAttachedToJni();
   }
 
+  /**
+   * Invoked when the {@link io.flutter.embedding.engine.FlutterEngine} that owns this
+   * {@link DartExecutor} detached from JNI.
+   *
+   * When detached from JNI, this {@link DartExecutor} stops handling 2-way communication to/from
+   * the Dart execution context.
+   */
   public void onDetachedFromJNI() {
-    isAttachedToJni = false;
     flutterJNI.setPlatformMessageHandler(null);
+    messenger.onDetachedFromJni();
   }
 
-  public boolean isApplicationRunning() {
+  /**
+   * Is this {@link DartExecutor} currently executing Dart code?
+   * @return true if Dart code is being executed, false otherwise
+   */
+  public boolean isExecutingDart() {
     return isApplicationRunning;
   }
 
-  public void runFromBundle(FlutterRunArguments args) {
-    if (args.bundlePath == null) {
-      throw new AssertionError("A bundlePath must be specified");
-    } else if (args.entrypoint == null) {
-      throw new AssertionError("An entrypoint must be specified");
+  /**
+   * Starts executing Dart code based on the given {@code dartEntrypoint}.
+   *
+   * See {@link DartEntrypoint} for configuration options.
+   *
+   * @param dartEntrypoint specifies which Dart function to run, and where to find it
+   */
+  public void executeDartEntrypoint(DartEntrypoint dartEntrypoint) {
+    if (isApplicationRunning) {
+      Log.w(TAG, "Attempted to run a DartExecutor that is already running.");
+      return;
     }
-    runFromBundleInternal(args.bundlePath, args.entrypoint, args.libraryPath, args.defaultPath);
+
+    flutterJNI.nativeRunBundleAndSnapshotFromLibrary(
+        nativeObjectReference,
+        dartEntrypoint.pathToPrimaryBundle,
+        dartEntrypoint.pathToFallbackBundle,
+        dartEntrypoint.dartEntrypointFunctionName,
+        null,
+        dartEntrypoint.androidAssetManager
+    );
+
+    isApplicationRunning = true;
   }
 
-  private void runFromBundleInternal(
-      String bundlePath,
-      String entrypoint,
-      String libraryPath,
-      String defaultPath
-  ) {
+  /**
+   * Starts executing Dart code based on the given {@code dartCallback}.
+   *
+   * See {@link DartCallback} for configuration options.
+   *
+   * @param dartCallback specifies which Dart callback to run, and where to find it
+   */
+  public void executeDartCallback(DartCallback dartCallback) {
     if (isApplicationRunning) {
-      throw new AssertionError("This Flutter engine instance is already running an application");
+      Log.w(TAG, "Attempted to run a DartExecutor that is already running.");
+      return;
```

Comment:
Why remove the exception? It should probably be IllegalStateException though.

Reply:
I didn't see any reason that calling this multiple times should blow up the app.  Do you?  If a developer really cares, he can query whether or not dart code is running, which is what we are now doing from FlutterFragment.

Reply:
Note to self: Turn it back into an exception and find out why it's a problem.  Then re-evaluate.